src/core/Akka.Streams/Dsl/Restart.cs

```diff
@@ -0,0 +1,490 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="Restart.cs" company="Akka.NET Project">
+//     Copyright (C) 2015-2017 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2017 Akka.NET project <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Akka.Pattern;
+using Akka.Streams.Stage;
+
+namespace Akka.Streams.Dsl
+{
+    /// <summary>
+    /// A RestartSource wraps a <see cref="Source"/> that gets restarted when it completes or fails.
+    /// They are useful for graphs that need to run for longer than the <see cref="Source"/> can necessarily guarantee it will, for
+    /// example, for <see cref="Source"/> streams that depend on a remote server that may crash or become partitioned. The
+    /// RestartSource ensures that the graph can continue running while the <see cref="Source"/> restarts.
+    /// </summary>
+    public static class RestartSource
+    {
+        /// <summary>
+        /// Wrap the given <see cref="Source"/> with a <see cref="Source"/> that will restart it when it fails or complete using an exponential
+        /// backoff.
+        /// This <see cref="Source"/> will never emit a complete or failure, since the completion or failure of the wrapped <see cref="Source"/>
+        /// is always handled by restarting it. The wrapped <see cref="Source"/> can however be cancelled by cancelling this <see cref="Source"/>.
+        /// When that happens, the wrapped <see cref="Source"/>, if currently running will be cancelled, and it will not be restarted.
+        /// This can be triggered simply by the downstream cancelling, or externally by introducing a <see cref="IKillSwitch"/> right
+        /// after this <see cref="Source"/> in the graph.
+        /// This uses the same exponential backoff algorithm as <see cref="Akka.Pattern.Backoff"/>.
+        /// </summary>
+        /// <param name="sourceFactory">A factory for producing the <see cref="Source"/> to wrap.</param>
+        /// <param name="minBackoff">Minimum (initial) duration until the child actor will started again, if it is terminated</param>
+        /// <param name="maxBackoff">The exponential back-off is capped to this duration</param>
+        /// <param name="randomFactor">After calculation of the exponential back-off an additional random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay. In order to skip this additional delay pass in `0`.</param>
+        public static Source<T, NotUsed> WithBackoff<T>(Func<Source<T, NotUsed>> sourceFactory, TimeSpan minBackoff, TimeSpan maxBackoff, double randomFactor) 
```

Comment:
`sourceFactory` should have a signature `Creator[Source[T, _]]`, I don't know how to do it on C#